File: app/src/main/java/com/udacity/google/popularmovies/util/FavouriteProvider.java
package com.udacity.google.popularmovies.util;

import android.content.ContentProvider;
import android.content.ContentValues;
import android.content.UriMatcher;
import android.database.Cursor;
import android.database.sqlite.SQLiteDatabase;
import android.database.sqlite.SQLiteOpenHelper;
import android.database.sqlite.SQLiteQueryBuilder;
import android.net.Uri;
import android.support.annotation.Nullable;

import java.util.HashMap;

/**
 * Created by Dell on 10/12/2015.
 */

public class FavouriteProvider extends ContentProvider {
    private FavouriteOpenHelper mOpenHelper;


    static final int uriCode = 1;
    static final UriMatcher uriMatcher;
    private static HashMap<String, String> values;
    static {
        uriMatcher = new UriMatcher(UriMatcher.NO_MATCH);
        uriMatcher.addURI(FavouriteContract.CONTENT_AUTHORITY, "favourite", uriCode);
        uriMatcher.addURI(FavouriteContract.CONTENT_AUTHORITY, "favourite/*", uriCode);
    }


    FavouriteDataSource dataSource;

    @Override
    public boolean onCreate() {
        dataSource = new FavouriteDataSource(getContext());

        dataSource.open();


        return true;
    }


    @Nullable
    @Override
    public Cursor query(Uri uri, String[] projection, String selection, String[] selectionArgs, String sortOrder) {


                SQLiteQueryBuilder qb = new SQLiteQueryBuilder();
                FavouriteDataSource dataSource = new FavouriteDataSource(getContext());

                qb.setTables(FavouriteContract.CONTENT_AUTHORITY);

                Cursor retCursor;
                switch (uriMatcher.match(uri)) {

                    case uriCode:
                        qb.setProjectionMap(values);
                        break;
                    default:
                        throw new UnsupportedOperationException("Unknown uri: " + uri);
                        }

        Cursor c = qb.query(dataSource.openHelper.getWritableDatabase(), projection, selection, selectionArgs, null,
                null, sortOrder);

        c.setNotificationUri(getContext().getContentResolver(), uri);
        return c;
    }

    @Nullable
    @Override
    public String getType(Uri uri) {
        switch (uriMatcher.match(uri)) {
            case uriCode:
               return FavouriteContract.FavouriteEntry.CONTENT_TYPE;
            default:
                throw new IllegalArgumentException("Unsupported URI: " + uri);
        }    }

    @Nullable
    @Override
    public Uri insert(Uri uri, ContentValues contentValues) {
        return null;
    }

    @Override
    public int delete(Uri uri, String s, String[] strings) {

        int count = 0;
        switch (uriMatcher.match(uri)) {
            case uriCode:
                count = FavouriteDataSource.database.delete(FavouriteContract.PATH_FAVOURITE, s, strings);

                break;
            default:
                throw new IllegalArgumentException("Unknown URI " + uri);
        }
        getContext().getContentResolver().notifyChange(uri, null);
        return count;
    }

    @Override
    public int update(Uri uri, ContentValues contentValues, String s, String[] strings) {
        int count = 0;
        switch (uriMatcher.match(uri)) {
            case uriCode:
                count = FavouriteDataSource.database.update(FavouriteContract.PATH_FAVOURITE, contentValues, s, strings);
                break;
            default:
                throw new IllegalArgumentException("Unknown URI " + uri);
        }
        getContext().getContentResolver().notifyChange(uri, null);
        return count;
    }






}

File: app/src/main/java/com/udacity/google/popularmovies/util/FavouriteProvider.kt
package com.udacity.google.popularmovies.util

import android.content.ContentProvider
import android.content.ContentValues
import android.content.UriMatcher
import android.database.Cursor
import android.database.sqlite.SQLiteDatabase
import android.database.sqlite.SQLiteOpenHelper
import android.database.sqlite.SQLiteQueryBuilder
import android.net.Uri

import java.util.HashMap

/**
 * Created by Dell on 10/12/2015.
 */

class FavouriteProvider : ContentProvider() {
    private val mOpenHelper: FavouriteOpenHelper? = null
    internal lateinit var dataSource: FavouriteDataSource

    override fun onCreate(): Boolean {
        dataSource = FavouriteDataSource(context!!)

        dataSource.open()


        return true
    }


    override fun query(uri: Uri, projection: Array<String>?, selection: String?, selectionArgs: Array<String>?, sortOrder: String?): Cursor? {


        val qb = SQLiteQueryBuilder()
        val dataSource = FavouriteDataSource(context!!)

        qb.tables = FavouriteContract.CONTENT_AUTHORITY

        val retCursor: Cursor
        when (uriMatcher.match(uri)) {

            uriCode -> qb.setProjectionMap(values)
            else -> throw UnsupportedOperationException("Unknown uri: $uri")
        }

        val c = qb.query(dataSource.openHelper.getWritableDatabase(), projection, selection, selectionArgs, null, null, sortOrder)

        c.setNotificationUri(context!!.contentResolver, uri)
        return c
    }

    override fun getType(uri: Uri): String? {
        when (uriMatcher.match(uri)) {
            uriCode -> return FavouriteContract.FavouriteEntry.CONTENT_TYPE
            else -> throw IllegalArgumentException("Unsupported URI: $uri")
        }
    }

    override fun insert(uri: Uri, contentValues: ContentValues?): Uri? {
        return null
    }

    override fun delete(uri: Uri, s: String?, strings: Array<String>?): Int {

        var count = 0
        when (uriMatcher.match(uri)) {
            uriCode -> count = FavouriteDataSource.database.delete(FavouriteContract.PATH_FAVOURITE, s, strings)
            else -> throw IllegalArgumentException("Unknown URI $uri")
        }
        context!!.contentResolver.notifyChange(uri, null)
        return count
    }

    override fun update(uri: Uri, contentValues: ContentValues?, s: String?, strings: Array<String>?): Int {
        var count = 0
        when (uriMatcher.match(uri)) {
            uriCode -> count = FavouriteDataSource.database.update(FavouriteContract.PATH_FAVOURITE, contentValues, s, strings)
            else -> throw IllegalArgumentException("Unknown URI $uri")
        }
        context!!.contentResolver.notifyChange(uri, null)
        return count
    }

    companion object {


        internal val uriCode = 1
        internal val uriMatcher: UriMatcher
        private val values: HashMap<String, String>? = null

        init {
            uriMatcher = UriMatcher(UriMatcher.NO_MATCH)
            uriMatcher.addURI(FavouriteContract.CONTENT_AUTHORITY, "favourite", uriCode)
            uriMatcher.addURI(FavouriteContract.CONTENT_AUTHORITY, "favourite/*", uriCode)
        }
    }


}
